Add linting with golangci-lint

jamietanna opened the pull request:
- add-lint
- C

Commit:
Add linting with golangci-lint

To give a bit more consistency across the project with the Go
ecosystem, as well as picking up on some common errors that new
contributions can bring, enable golangci-lint, with default
configuration.

Commit:
Lint: remove unused variable

Commit:
Lint: Check return values

Right now, we can ignore the return values, as we're hardcoding the
response of an `err` to always return, which isn't ideal.

Commit:
Lint: Remove unused unexported variable

Commit:
Lint: Use consistent Go-style package names

Commit:
Lint: `unconvert`

Commit:
Lint: Exclude unexported fields for `structcheck`

As we've added them on purpose, to show that they don't get
(un)marshalled.

Commit:
Lint: Ignore findings on string-based keys for `Context`

As it's probably a bit late to change this, in case it changes expected
behaviour, via SA1029.

Commit:
Lint: Use backtick'd strings for regexes

Via S1007

Commit:
Lint: Don't use deprecated `strings.Title`

Via SA1019

Commit:
Lint: remove ineffectual assignments

Commit:
Lint: Ensure we check errors

Or, least show that we've considered whether to ignore them, which we do
to follow existing style of the file to ignore everything.

Commit:
Lint: Use `fmt.Errorf` instead of `errors.New(fmt.Sprintf)`

Via S1028

Commit:
Lint: remove warning for `ineffectual`

As we'll come back to it later.

pkg/runtime/deepobject.go:245:5: ineffectual assignment to tm (ineffassign)
                                tm, err = time.Parse(types.DateFormat, pathValues.value)

Commit:
Lint: Simplify `make` for a slice

Commit:
Lint: Ignore marshalling warning with `reflect.Value`

pkg/runtime/bindform.go:102:34: SA9005: struct type 'reflect.Value' doesn't have any exported fields, nor custom marshaling (staticcheck)
                                if data, err := json.Marshal(field); err != nil {

jamietanna commented:
@deepmap-marcinr mind having a look? I need to clean up the commit history (hence this still being a draft) but would be good to get this through to start giving us a bit more control + confidence in our code style, as well as closing #58.

jamietanna commented on pkg/gin-middleware/oapi_validate.go:
```diff
@@ -122,12 +122,12 @@ func ValidateRequestFromContext(c *gin.Context, router routers.Router, options *
 	// Pass the gin context into the request validator, so that any callbacks
 	// which it invokes make it available.
-	requestContext := context.WithValue(context.Background(), GinContextKey, c)
+	requestContext := context.WithValue(context.Background(), GinContextKey, c) //nolint:staticcheck
```
SA1029: should not use built-in type string as key for value; define your own type to avoid collisions (staticcheck)

jamietanna commented on pkg/runtime/bindform.go:
```diff
@@ -99,7 +99,7 @@ func MarshalForm(ptr interface{}, encodings map[string]RequestBodyEncoding) (url
 		if encoding, ok := encodings[tag]; ok && encoding.ContentType != "" {
 			if strings.HasPrefix(encoding.ContentType, jsonContentType) {
-				if data, err := json.Marshal(field); err != nil {
+				if data, err := json.Marshal(field); err != nil { //nolint:staticcheck
```
`SA9005: struct type 'reflect.Value' doesn't have any exported fields, nor custom marshaling`

jamietanna commented on pkg/runtime/deepobject.go:
```diff
@@ -242,7 +242,7 @@ func assignPathValues(dst interface{}, pathValues fieldOrValue) error {
 			if err != nil {
 				// Fall back to parsing it as a date.
-				tm, err = time.Parse(types.DateFormat, pathValues.value)
+				tm, err = time.Parse(types.DateFormat, pathValues.value) //nolint:ineffassign,staticcheck
```
`ineffectual assignment to tm (ineffassign)`

deepmap-marcinr commented:
I rebased your branch, and fixed up a bunch of merge conflicts, however, you're making breaking API changes due to initialisms - people depending on the testutils or runtime packages will see code break. Our hands are kinda tied there until a breaking release.

jamietanna commented:
> I rebased your branch, and fixed up a bunch of merge conflicts, however, you're making breaking API changes due to initialisms - people depending on the testutils or runtime packages will see code break. Our hands are kinda tied there until a breaking release.

Ah that's fair, I'll ignore the initialisms check, for now then 👍

stevenh commented:
Hi @jamietanna I added `golangci-lint` separately in https://github.com/deepmap/oapi-codegen/pull/845 see if it does what you need it to?

stevenh commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,20 @@
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version-file: 'go.mod'
```
This is the default so not needed
```suggestion
```

stevenh commented on internal/test/issues/issue-312/issue_test.go:
```diff
@@ -26,7 +26,8 @@ func TestClient_WhenPathHasColon_RequestHasCorrectPath(t *testing.T) {
 	})
 
-	client.ValidatePetsWithResponse(context.Background(), ValidatePetsJSONRequestBody{
+	// TODO: this err should be checked, but it defaults to returning the "something went wrong", so we should refactor accordingly
```
We should just check the errors, its easy enough, also check the errors for NewClient above

stevenh commented on pkg/codegen/prune.go:
```diff
@@ -28,14 +28,14 @@ func walkSwagger(swagger *openapi3.T, doFn func(RefWrapper) (bool, error)) error
 	for _, p := range swagger.Paths {
 		for _, param := range p.Parameters {
-			walkParameterRef(param, doFn)
+			_ = walkParameterRef(param, doFn)
```
This is bad, you should never do this to quash the lint error, just check the error, lots more below.

jamietanna replied:
Although yes, I 100% agree, the file so far has a lot of these, and I'd prefer us to come back and sort them

stevenh replied:
I see a two options:
1. Disable `errcheck` linter in `.golangci-lint.yaml` for now.
2. Use https://github.com/deepmap/oapi-codegen/pull/845 which already fixes all of these.

stevenh reviewed (changes requested):
Few nits but the the main issue is just using `_ = function()` to quell error not checked, we should never do that.

stevenh commented:
Great to see this merged!

I've rebased [my golangci-lint](https://github.com/deepmap/oapi-codegen/pull/871) changes on top of it, hopefully we can get that in too.
